### PR TITLE
LOOP-488 Bolus progress and quick cancellation

### PR DESCRIPTION
**Pull request by ps2**

### CODE REVIEW

File: Loop/View Controllers/StatusTableViewController.swift [hunk 1]
```diff
@@ -16,24 +16,19 @@ import SwiftCharts
...
 private extension RefreshContext {
     static let all: Set<RefreshContext> = [.status, .glucose, .insulin, .carbs, .targets]
 }
 
-
 final class StatusTableViewController: ChartsTableViewController {
 
     private let log = OSLog(category: "StatusTableViewController")
 
+    lazy var quantityFormatter: QuantityFormatter = {
```
Comment (mpangburn) [inline Loop/View Controllers/StatusTableViewController.swift]: `lazy var quantityFormatter = QuantityFormatter()`

File: Loop/View Controllers/StatusTableViewController.swift [hunk 2]
```diff
@@ -713,17 +735,24 @@ final class StatusTableViewController: ChartsTableViewController {
 
             return cell
         case .status:
-            let cell = tableView.dequeueReusableCell(withIdentifier: TitleSubtitleTableViewCell.className, for: indexPath) as! TitleSubtitleTableViewCell
-            cell.selectionStyle = .none
+
+            let getTitleSubtitleCell: () -> TitleSubtitleTableViewCell = {
```
Comment (mpangburn) [inline Loop/View Controllers/StatusTableViewController.swift]: Clearer as a `func`

File: Loop/Views/BolusProgressTableViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import HealthKit
+import MKRingProgressView
+
+public class BolusProgressTableViewCell: UITableViewCell {
+
+    @IBOutlet weak var progressLabel: UILabel!
+
+    @IBOutlet weak var stopSquare: UIView! {
+        didSet {
+            stopSquare.layer.cornerRadius = 3
```
Comment (mpangburn) [inline Loop/Views/BolusProgressTableViewCell.swift]: I think this corner radius is _slightly_ too large. I played briefly with changing it and found that a value of 2 preserves the square shape more clearly.

File: Loop/Views/BolusProgressTableViewCell.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private func updateProgressColor() {
+        progressIndicator.startColor = tintColor
+        progressIndicator.endColor = tintColor
+        stopSquare.backgroundColor = tintColor
+    }
+
+    private func updateProgress() {
+        if let totalUnits = totalUnits, let unit = unit {
```
Comment (mpangburn) [inline Loop/Views/BolusProgressTableViewCell.swift]: Nit: `guard` here

File: Loop/Views/TitleSubtitleTableViewCell.swift [hunk 5]
```diff
@@ -12,6 +12,8 @@ class TitleSubtitleTableViewCell: UITableViewCell {
 
     @IBOutlet weak var titleLabel: UILabel!
 
+    @IBOutlet weak var progressIndicator: UIProgressView!
```
Comment (mpangburn) [inline Loop/Views/TitleSubtitleTableViewCell.swift]: Is this a remnant from a previous UI implementation?

**Review by mpangburn — APPROVED** [review]
Looks really nice. Just a couple UI fixes:
- The status row hides before the ring closing animation finishes for a completed bolus.
- When another bolus starts after one has completed, the full ring's value is visibly unwound when the status row appears. The ring should be reset before the user sees it again.

File: Loop/Views/BolusProgressTableViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                }
+            } else {
+                progressLabel.text = String(format: NSLocalizedString("Bolusing %1$@", comment: "The format string for bolus in progress showing total volume. (1: total volume)"), totalUnitsString)
+            }
+        }
+    }
+
+    override public func prepareForReuse() {
+        super.prepareForReuse()
+        progressIndicator.progress = 0
```
Comment (darinkrauss) [inline Loop/Views/BolusProgressTableViewCell.swift:106]: `progressLabel.text = ""`?

**Review by darinkrauss — APPROVED** [review]
LGTM 🚀

**Review by mpangburn — APPROVED** [review]
Nothing to add since the last review. 🚀

**Review by darinkrauss — APPROVED** [review]
LGTM (again) 🏎

**Pull request merged**